### ISSUE TITLE
Object graph comparison is broken when type implements IEnumerable?

### ISSUE THREAD

**Issue by jstachowiak** [fluentassertions/fluentassertions#860]
Hello!

When a type is a collection the comparison seems to ignore all other type members. Is this by design or am I missing here something? What is the best workaround for now? I'm using version `5.3.2`.

```csharp
public class FluentAssertionsTests
{
    public class CollectionA : List<ObjectA>
    {
        public int Y { get; set; }
    }

    public class ObjectA
    {
        public int X { get; set; }
    }

    [Fact]
    public void ThisFailsAsExpected()
    {
        var a1 = new ObjectA();
        var a2 = new ObjectA();

        a1.X = 1;
        a2.X = 2;

        a1.Should().BeEquivalentTo(a2);
    }

    [Fact]
    public void ThisShouldFailButSucceeds()
    {
        var a1 = new CollectionA {new ObjectA {X = 1}};
        var a2 = new CollectionA {new ObjectA {X = 1}};

        a1.Y = 1;
        a2.Y = 2;

        a1.Should().BeEquivalentTo(a2);
    }
}
```

Thanks :)

**Comment by frankschierle** [comment]
Im currently migrating from 4.0.0 to 5.4.1 and stumpled accross an issue that is probably related to this one.

The following code doesn't throw an exception in version 4.0.0:

```csharp
List<Type> types1 = new List<Type> { typeof(int), typeof(string) };
IEnumerable types2 = new List<Type> { typeof(string), typeof(int) };

types1.Should().BeEquivalentTo(types2);
```

In version 5.4.1 it throws an exception. When using `IEnumerable<Type>` no exception is thrown.

**Comment by jnyrup** [comment]
@frankschierle From your code snippet it looks like you need to distinguish between [Compile-time types vs. run-time types](https://fluentassertions.com/documentation/#compile-time-types-vs-run-time-types).
To use runtime-time types instead of compile-time types, write `types1.Should().BeEquivalentTo(types2, options => options.RespectingRuntimeTypes())`.

If that is not the problem, please post the stack trace.

**Comment by frankschierle** [comment]
@jnyrup The compiler won't allow me to specify options, if  `expectation` is an non-generic `IEnumerable`.

```csharp
List<Type> types1 = new List<Type> { typeof(int), typeof(string) };
IEnumerable types2 = new List<Type> { typeof(string), typeof(int) };

types1.Should().BeEquivalentTo(types2, options => options.RespectingRuntimeTypes());
```

Compiler error:
```
CS1660	Cannot convert lambda expression to type 'object[]' because it is not a delegate type
```

The overload of `BeEquivalentTo()` that accepts options, requires a generic `IEnumerable<T>`.

**Comment by jnyrup** [comment]
@frankschierle I see.
You can use `types1.Should().BeEquivalentTo(types2 as IEnumerable<object>, options => options.RespectingRuntimeTypes());`
If that's not an option, please open a new issue as your issue is different from this.

**Comment by frankschierle** [comment]
@jnyrup That's almost exactly the workaround I'm currently using.

I opened a new issue (#881), because I think that's a regression.

**Comment by dennisdoomen** [comment]
Is this issue still relevant?

**Comment by danielmpetrov** [comment]
Hi, I'm new to this project, and I'm trying to investigate how this bug could be solved.

Has anyone attempted this, or has tips/guidance on how to tackle the problem?

**Comment by dennisdoomen** [comment]
It's more of a functional issue. When an object implements `IEnumerable`, the enumerable-focused equivalency steps are picked up [first](https://github.com/fluentassertions/fluentassertions/blob/master/Src/FluentAssertions/EquivalencyStepCollection.cs#L113). We _could_ allow multiple equivalency steps pick up the same object, but I'm not sure if that is going to cause side effects.

**Comment by FabianNitsche** [comment]
Could you please summarize what to fix? Is it that
```cs
class A : IEnumerable<B>
{ 
    public C C { get; } 
    ... 
}
class D : IEnumerable<B> 
{ 
    public C C { get; } 
    ... 
}
someA.Should().BeEquivalentTo(someD)
```
is not respecting the C property in the comparison?

**Comment by dennisdoomen** [comment]
> is not respecting the C property in the comparison?

Correct. Since both `A` and `D` implement `IEnumerable<T>`, FA will only compare the objects in the collection and skip `C`.

**Comment by FabianNitsche** [comment]
I have been looking into this, and I think the problem is, that the enumerable classes are treated differently from the rest.

A structural comparison is by default not possible, since the BeEquivalentTo() method in CollectionAssertions is defined as:
```cs
public AndConstraint<TAssertions> BeEquivalentTo<TExpectation>(IEnumerable<TExpectation> expectation, ...
```
where in ObjectAssertions it is defined as
```cs
public AndConstraint<ObjectAssertions> BeEquivalentTo<TExpectation>(TExpectation expectation, ...
```
This already leads to the wrong CompileTimeType for the CollectionAssertions. The former method should probably read something like:
```cs
public AndConstraint<TAssertions> BeEquivalentTo<TExpectation>(TExpectation expectation, ...
    where TExpectation : IEnumerable<TElement>
```
with TElement being defined by the enclosing class.

Due to the "wrong" CompileTimeType a structural comparison in the StructuralEqualityEquivalencyStep misses all properties.

But there is another problem preventing even getting to the structural comparison in the list of EquivalencySteps. Again it is because enumerables are treated differently: While other equivalency steps check for the exact type, the collection equivalency steps check that the compile type implements the interface and are happy to take the lead, thus skipping all remaining steps (including the structural test, that would see the property difference). 

I think to fix that it is probably better to treat the implementation of the enumerable interface in the same way members are treated. 

Basically, the collection equivalency steps check for exact type and not interfaces and the structural equivalency step treats the implementation of the enumerable interface as a member to be compared. (so it is not done at root level but as a sublevel)

I do not think there is an easy fix here without making the code much more complicated. If you are generally OK with such major changes (while not adding complexity), I would give it a try. So that we have something more concrete to discuss on. No promisses though, that I will find a sensible solution... :D

**Comment by dennisdoomen** [comment]
> Basically, the collection equivalency steps check for exact type and not interfaces and the structural equivalency step treats the implementation of the enumerable interface as a member to be compared. (so it is not done at root level but as a sublevel)

This is what I'm struggling with. A type implementing `IEnumerable<T>` will not have any public properties or fields that you can use to iterate the collection.

**Comment by FabianNitsche** [comment]
Yes, but I would suggest to put the special treatment of enumerables/collections/... there. I would give it a shot in implementing such a solution, so that we have something concrete to discuss further. I am totally OK if you eventually dismiss the whole thing. Still I need to know if you are willing at all to incorporate such "larger" changes, or if you want to have a less intrusive fix.

**Comment by dennisdoomen** [comment]
Sure, because right now, I don't really now how to solve this, neither technically nor functionally.

**Comment by FabianNitsche** [comment]
So I have a first implementation, treating IEnumerable implementations similar to properties or fields. And than I think I now understand the problem much better. The basic problem is, that BeEquivalentTo is behaving differently for objects and Collections. In the first case the object graph is compared so if a property is missing the comparison will complain, where the latter does only look at the elements of the enumerable.

Meaning, if I compare two objects that do not implement any collection interface but one has the property Count and the other Length, BeEquivalentTo will complain, because the two objects are structurally different. However, if I compare e.g. an empty List, that has a Count property to an empty Array that has a Length property BeEquivalentTo will not complain.

This basically originates, that any object that implements a collection interface is treated differently. And this comes from a priority list called EquivalencyStepsCollection.GetDefaultSteps, where collections are tested prior to structural equivalency.

Now as I said it is possible to regard the enumerable part of the object structure as something similar to a property and remove all the dedicated collection equivalency steps from the list. This also allows types to implement multiple enumerables. However, this leads to a breaking change, since now Arrays are not equivalent to Lists anymore, because the one calls it Count and the other one Length.

In summary, I think BeEquivalent behaves way to different when called with objects implementing collection interfaces than it behaves called with objects not implementing collection interfaces. So the name should be different. A way to keep the breaking change minimal could be to ignore the properties of framework collections.

**Comment by dennisdoomen** [comment]
It's even more complicated because FA treats two collections as equivalent if they contain equivalent objects in any order. Maybe we need something like we have for value types (`ComparingByValue` and `ComparingByMembers`) to override the behavior for a type that is treated as a collection.

**Comment by FabianNitsche** [comment]
So would you accept a breaking change? So that BeEquivalentTo now looks for structural equivalency also for collections? 

Maybe add an option to allow different ordering for collections in the structural equivalency and dedicated methods that only look at the elements of the ienumerable?

The special treatment of the Collection equivalency steps would then go away.

**Comment by dennisdoomen** [comment]
No, but if we use the option `ComparingByMembers<MyCollectionType>`, we could allow a user to tell FA to treat that type as an ordinary type with properties.

**Comment by FabianNitsche** [comment]
OK, I understand. But do you agree, that BeEquivalentTo behaves unexpected with collections? So this would be like a mild, slightly inconsistent fix?

**Comment by dennisdoomen** [comment]
No, not really. I designed it like this. However, I do acknowledge that you sometimes want different behavior for a collection type that has other properties. So for those I propose to honor the `ComparingByMembers` method.

**Comment by FabianNitsche** [comment]
OK. Thanks for the info. So probably the Collection EquivalencySteps should respect that option then. And if ComparingByMembers is set then a List of elements is not equivalent to an array of the same elements, right?

**Comment by dennisdoomen** [comment]
Well, if the collection type has a property that itself is a collection, for that property, the `EnumerableEquivalencyStep` will still kick in. But the collection type itself will still be treated as a complex type.

**Comment by FabianNitsche** [comment]
Lets say we have two independent types with equal implementation EnumerableWithPropertyA and EnumerableWithPropertyB the first TypeParameter gives the property type the second is the type of the elements of IEnumerable<T> these types are implementing.
Additionally we also define a reference to customCollectionA where the compile time type is object.
``` C#
var customCollectionA = new EnumerableWithPropertyA<string, int>("Some String", 1, 2, 3);
var customCollectionB = new EnumerableWithPropertyB<string, int>("Some other string", 3, 2, 1);
object customCollectionAObject = customCollectionA;
```
Now if we use the latter for assertion, the compile time type of the expectation is  EnumerableWithPropertyB<string, int>, because the method signature is `BeEquivalentTo<TExpectation>(TExpectation expectation, ...)` and the compile time type of the expectation is `TExpectation`. So I have a fix that lets the following two tests pass:
``` C#
customCollectionAObject.Should().BeEquivalentTo(customCollectionB);

Action usingObjectAssertions = () => customCollectionAObject.Should().BeEquivalentTo(customCollectionB, options => options.ComparingByMembers<EnumerableWithPropertyB<string, int>>());
usingObjectAssertions.Should().Throw<XunitException>();
```
From my point of view, this is what we discussed a solution should look like.
_BTW: Instead of passing `EnumerableWithPropertyB<string, int>` as the type parameter for `ComparingByMembers` you can also pass `IEnumerable<int>`, `IEnumerable`, or `object`. But you cannot pass `EnumerableWithPropertyA<string, int>`. I think this is also a somewhat strange behavior, because from the signature and name of the method I would expect, that the checks are only done on the inheritance level of the type specified. This is a side note, that we may have to discuss later._ 

Now the main problem I have is the following, and here I need your input. If I do not change the compile time type of `customCollectionA` to `object` the CollectionAssertions are used instead. And there the assertion method signature is `BeEquivalentTo<TExpectation>(IEnumerable<TExpectation> expectation, ...` and so we loose the compile time type of the expectation and consequently we cannot compare the type specified in `ComparingByMembers`

Thus the second test here fails:
``` C#
customCollectionA.Should().BeEquivalentTo(customCollectionB);

// this fails:
Action usingGenericCollectionAssertions = () => customCollectionA.Should().BeEquivalentTo(customCollectionB, options => options.ComparingByMembers<EnumerableWithPropertyB<string, int>>());
usingGenericCollectionAssertions.Should().Throw<XunitException>();
```

To fix this one would need to change the whole signatures of the CollectionAssertions I gave it a try once but this has rippling changes all over the code. A fix could be to add another type parameter: `CollectionAssertions.BeEquivalentTo<TExpectation, TElement>(TExpectation expectation, ...) where TExpectation : IEnumerable<TElement>` this has actually the advantage, that one could specify individual options for the collection and element type.

Sorry for the lengthy post. The main question is: Is the test behavior as you would expect it and if not what direction would you like to take it further.

**Comment by dennisdoomen** [comment]
> I think this is also a somewhat strange behavior, because from the signature and name of the method I would expect, that the checks are only done on the inheritance level of the type specified. This is a side note, that we may have to discuss later.

I expect it to honor the _is a_ concept. So if it _is_ that specific or inherits/implements from it. 

> Now the main problem I have is the following, and here I need your input. If I do not change the compile time type of customCollectionA to object the CollectionAssertions are used instead. And there the assertion method signature is BeEquivalentTo<TExpectation>(IEnumerable<TExpectation> expectation, ... and so we loose the compile time type of the expectation and consequently we cannot compare the type specified in ComparingByMembers

I that case you're even more explicit that you're comparing two collections. I would not even consider the fact that my collection type may have other properties. But you can still use the `ComparingByMember` option since internally it's still using the same equivalency engine.

**Comment by FabianNitsche** [comment]
OK, but to get the exact same behavior for the equivalency engine, we need to have the compile time type. And as I said this is lost for `CollectionAssertions.BeEquivalentTo<TExpectation>(IEnumerable<TExpectation> expectation, ...). And we cannot use the runtime type, because then we would not use the same equivalency rules.
So how do we pass the compile time type?

**Comment by dennisdoomen** [comment]
I don't get the compile time type problem.

**Comment by FabianNitsche** [comment]
Maybe I am phrasing it wrong. When you look at the following method signatures:
`ObjectAssertions.BeEquivalentTo<TExpectation>(TExpectation expectation, ...)`
and
`CollectionAssertions.BeEquivalentTo<TExpectation>(IEnumerable<TExpectation> expectation, ...)`
In the first case I get the compile time type of the expectation and in the second case I get the compile time type of the element of the expectation that implements `IEnumerable<T>`. The compile time type of the expectation that implements IEnumerable<T> is lost. So I cannot check the compile time type against the type that was passed to the options by `ComparingByMembers`. Because the passed compile time type is the one of the element of the expectation and of the expectation itself.

**Comment by dennisdoomen** [comment]
Why not? Internally FA uses Reflection anyhow. And the second example internally does this:

`CompileTimeType = typeof(IEnumerable<TExpectation>),`

**Comment by FabianNitsche** [comment]
> Why not? Internally FA uses Reflection anyhow. And the second example internally does this:
`CompileTimeType = typeof(IEnumerable<TExpectation>),`

And this is not good. Because this shows, how differently the treatment of a type that implements `IEnumerable<T>` and a type that does not implement it is.

Maybe the following makes the problem clearer. Look at this signature:
```C#
BeEquivalentTo<TExpectation>(IEnumerable<TExpectation> expectation,
            Func<EquivalencyAssertionOptions<TExpectation>, EquivalencyAssertionOptions<TExpectation>> config, ...)
```
I can only specify options for the element type of the IEnumerable not for the type that implements the IEnumerable. So when we look at the example I gave before:
```C#
customCollectionA.Should().BeEquivalentTo(customCollectionB, options => options.ComparingByMembers<EnumerableWithPropertyB<string, int>>());
```
I am telling the equivalency engine to look if the IEnumerable element type (in this case `int`) is or inherits from `EnumerableWithPropertyB<string, int>`. There is no way to specify this for the type that implements the `IEnumerable<TExpectation>`.

**Comment by dennisdoomen** [comment]
I think you're wrong. The last example would do exactly what you're expecting. It current tells the engine that it should compare any instances of `EnumerableWithPropertyB<string, int>` by its members instead of it's `Equals` method. We could add this same logic to the enumerable equivalency step. If it's marked as "compare-by-members", don't treat it as an enumerable.

**Comment by FabianNitsche** [comment]
Hm... I do not know if this is just a misunderstanding or if we are expecting different things here.
So I am trying to go back a little. (I am a newbie so tell me if I am going to far and should just stop, your library helped me a lot in my work and I just wanted to give back, but maybe I am not fit for this.)

In the example above the **subject** (the instance `Should` is called on) is always the same instance but once its compile time type is `EnumerableWithPropertyA<string, int>` and once it is `object`.

I do expect both calls to do **exactly** the same.

**Comment by dennisdoomen** [comment]
I might be wrong myself or misunderstand your problem, but the compile-type type _does_ make a difference. See also https://fluentassertions.com/objectgraphs/#compile-time-types-vs-run-time-types

**Comment by FabianNitsche** [comment]
OK I think I get our misunderstanding. The compile and runtime type is determined by the expectation and not by the subject. You can see this in the implementation of EquivalencyValidationContext:
```C#
       /// <summary>
        /// Gets the run-time type of the current expectation object.
        /// </summary>
        public Type RuntimeType
        {
            get
            {
                if (Expectation != null)
                {
                    return Expectation.GetType();
                }

                if (SelectedMemberInfo != null)
                {
                    return SelectedMemberInfo.MemberType;
                }

                return CompileTimeType;
            }
        }
```
And when the CompileTimeType is set in the Context. E.g. ObjectAssertions:
```C#
public AndConstraint<ObjectAssertions> BeEquivalentTo<TExpectation>(TExpectation expectation,
            Func<EquivalencyAssertionOptions<TExpectation>, EquivalencyAssertionOptions<TExpectation>> config, string because = "",
            params object[] becauseArgs)
        {
            Guard.ThrowIfArgumentIsNull(config, nameof(config));

            EquivalencyAssertionOptions<TExpectation> options = config(AssertionOptions.CloneDefaults<TExpectation>());

            var context = new EquivalencyValidationContext
            {
                Subject = Subject,
                Expectation = expectation,
                CompileTimeType = typeof(TExpectation),
                Because = because,
                BecauseArgs = becauseArgs,
                Tracer = options.TraceWriter
            };

            var equivalencyValidator = new EquivalencyValidator(options);
            equivalencyValidator.AssertEquality(context);

            return new AndConstraint<ObjectAssertions>(this);
        }
```
But if you look at the previous example the runtime and compilation type of the expectation is the same. **Only the compile time type of the subject is different.** And it is unexpected that this difference changes the behavior.

**Comment by FabianNitsche** [comment]
Could you please comment?

**Comment by dennisdoomen** [comment]
> But if you look at the previous example the runtime and compilation type of the expectation is the same. Only the compile time type of the subject is different. And it is unexpected that this difference changes the behavior.

Sorry, I'm really struggling to understand your point and how that relates to our original discussion on introducing `ComparingByMembers` for enumerables. _Which_ previous example? 

PS. It really helps if you use links, punctuation and such. Or maybe we should move the discussion to [Slack](https://fluentassertionsslack.herokuapp.com/)

**Comment by FabianNitsche** [comment]
I am sorry, one should not write such things in a hurry.

**Comment by dennisdoomen** [comment]
As there was no further explanation, closing this because of inactivity
